### ISSUE TITLE
Add trait/type name to the span name on trait methods that use instrument

### ISSUE THREAD

**Issue by tiagolobocastro** [tokio-rs/tracing#1939]
### Crates
tracing

### Motivation

When we have a trait, some trait methods may have less descriptive names on their own, example:
```rust
#[async_trait]
impl trait Pool for PoolXpto {
#[instrument]
async fn create(&self) { ... }
}
```
In some cases, you might want a more specific name for the span, example "Pool::create" or "Pool::PoolXpto::create". In this case I ended up changing each instrument, example:
```rust
#[async_trait]
impl trait Pool for PoolXpto {
#[instrument(name = "Pool::PoolXpto::create")]
async fn create(&self) { ... }
}
```

This process of manually changing the name is a bit tedious and error prone. It would be nice if we could somehow do this automatically.

### Proposal

1. Maybe we can have a post fix that gets automatically added to the name? Example:
```rust
#[async_trait]
impl trait Pool for PoolXpto {
#[instrument(postfix = "Pool::PoolXpto")] // -> span name would be "Pool::PoolXpto::create"
async fn create(&self) { ... }
}
```
2. Can we add some information to the trait that the instrument might be able to make use of? Example:
```rust
#[async_trait]
#[tracing(trait_name)]
impl trait Pool for PoolXpto {
#[instrument] // => this would then create a span with name "Pool::PoolXpto::create"
async fn create(&self) { ... }
}
```
Maybe we could have a trait that gets implemented with "tracing" and returns the postfix name which then instrument uses. Not sure how feasible this is though. Maybe something like:
```rust
trait TracingPostfix<Trait: ?Sized> {
fn postfix(&self) -> &static str;
}
impl trait TracingPostfix<dyn PoolXpto> for Pool {
fn postfix(&self) -> &'static str { "Pool::PoolXpto" }
}
```

### Alternatives

Simply add the names one by one on each trait method and be sure to set them correctly.

**Comment by hawkw** [comment]
See also https://github.com/tokio-rs/tracing/issues/1123.

A somewhat simpler API surface for a feature like this is just to allow placing the `#[instrument]` attribute on `impl` blocks, and have it modify any `#[instrument]` attribute inside the `impl` block to pass the of the type in the `impl` block as part of the `name` argument to the `#[instrumen]` attribute.

For example:

```rust
#[instrument] 
impl Foo {
    #[instrument] 
    fn bar(&self) {
        // ...
    }
}
```
would modify the attribute on the function to 
````rust
    #[instrument(name = "Foo::bar)] 
    fn bar(&self) {
        // ...
    }
```

Similarly, on a trait impl block:

```rust
#[instrument] 
impl Baz for Foo {
    #[instrument] 
    fn baz(&self) {
        // ...
    }
}
```
would modify the attribute on the function to 
````rust
    #[instrument(name = "<Foo as Baz>::baz")] 
    fn baz(&self) {
        // ...
    }
```

What do you think?

**Comment by tiagolobocastro** [comment]
Yeah that sounds good and it's pretty tidy as it's using the same macro.
Would it be worth adding an argument to the `instrument` on the `impl` blocks, to make it clear what it's doing? Otherwise could it lead someone to believe it's automatically adding `instrument` to all methods within the `impl`? Though this can also be address with documentation anyway.
